### PR TITLE
[light] Add ``initial_state`` configuration

### PR DESCRIPTION
**Pull request by jesserockz**
# What does this implement/fix?

This adds a new `initial_state` configuration option for all lights to initialise their state to a desired state if not restored from preferences.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#4331

## Test Environment

- [ ] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
light:
  - platform: ...
    ...
    initial_state:
      color_mode: rgb
      red: 100%
      green: 0%
      blue: 100%
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/light/light_state.cpp [hunk 1]
```diff
@@ -1,5 +1,5 @@
-#include "esphome/core/log.h"
 #include "light_state.h"
+#include "esphome/core/log.h"
```
Comment (kbx81) [inline esphome/components/light/light_state.cpp:1]: Moving this include seems weird. haha

File: esphome/components/light/__init__.py [hunk 2]
```diff
@@ -145,6 +151,32 @@ async def setup_light_core_(light_var, output_var, config):
...
 
+    if (initial_state_config := config.get(CONF_INITIAL_STATE)) is not None:
+        initial_state = cg.StructInitializer(
+            LightStateRTCState,
+            (
+                "color_mode",
+                initial_state_config.get(CONF_COLOR_MODE, ColorMode.UNKNOWN),
+            ),
+            ("state", initial_state_config.get(CONF_STATE, False)),
+            ("brightness", initial_state_config.get("brightness", 1.0)),
```
Comment (kbx81) [inline esphome/components/light/__init__.py]: Curious, why not import the constants for these and use them, instead? Like this:
```suggestion
            ("brightness", initial_state_config.get(CONF_BRIGHTNESS, 1.0)),
```

Comment (jesserockz) [reply esphome/components/light/__init__.py]: That's copilot for ya. Will fix them

**Review by kbx81 — DISMISSED** [review]

File: esphome/components/light/__init__.py [hunk 3]
```diff
@@ -145,6 +158,32 @@ async def setup_light_core_(light_var, output_var, config):
...
+            ("blue", initial_state_config.get(CONF_BLUE, 1.0)),
+            ("white", initial_state_config.get(CONF_WHITE, 1.0)),
+            ("color_temp", initial_state_config.get(CONF_COLOR_TEMPERATURE, 1.0)),
+            (
+                "cold_white",
+                initial_state_config.get(CONF_COLD_WHITE_COLOR_TEMPERATURE, 1.0),
+            ),
+            (
+                "warm_white",
+                initial_state_config.get(CONF_WARM_WHITE_COLOR_TEMPERATURE, 1.0),
```
Comment (kbx81) [inline esphome/components/light/__init__.py]: Noticed as I'm working on docs for this -- here we use `CONF_COLD_WHITE_COLOR_TEMPERATURE` and `CONF_WARM_WHITE_COLOR_TEMPERATURE` but in `LIGHT_STATE_SCHEMA`  we use `CONF_COLD_WHITE` and `CONF_WARM_WHITE`...

**Review by kbx81 — APPROVED** [review]

**Pull request merged**